### PR TITLE
[mlir][tensor] Enhance SimplifyUnPackToCollapseShape for unit dim cases.

### PR DESCRIPTION
**Pull request by hanhanW**
Progress on https://github.com/openxla/iree/issues/16181

### CODE REVIEW

**Comment by hanhanW** [comment]
This depends on https://github.com/llvm/llvm-project/pull/79247, please review the last commit

**Comment by hanhanW** [comment]
I rebased it to ToT, and drop the parent commit. Please take a look, thanks!

File: mlir/lib/Dialect/Tensor/Transforms/PackAndUnpackPatterns.cpp [hunk 1]
```diff
@@ -134,9 +135,36 @@ struct SimplifyUnPackToCollapseShape : public OpRewritePattern<UnPackOp> {
...
     if (dimsPos.size() != 1 || (dimsPos[0] + 1 != destType.getRank())) {
       return rewriter.notifyMatchFailure(
-          unpackOp, "expects unpacking at the innermost dimension");
+          unpackOp, "expects unpacking on the innermost dimension");
     }
 
+    return success();
+  }
+
+  /// Returns success() if it unpacks a 1D or 2D source operand to a 1D
```
Comment (chelini) [inline mlir/lib/Dialect/Tensor/Transforms/PackAndUnpackPatterns.cpp:144]: nit: the comment does not reflect implementation. I don't see why the source must be a 1d or a 2d tensor.

File: mlir/lib/Dialect/Tensor/Transforms/PackAndUnpackPatterns.cpp [hunk 2]
```diff
@@ -134,9 +135,36 @@ struct SimplifyUnPackToCollapseShape : public OpRewritePattern<UnPackOp> {
...
+  /// destination.
+  LogicalResult isUnpackTo1DDest(RewriterBase &rewriter,
```
Comment (chelini) [inline mlir/lib/Dialect/Tensor/Transforms/PackAndUnpackPatterns.cpp:146]: this is identical to `isPack1DSrc`. Can we share some code?

Comment (hanhanW) [reply mlir/lib/Dialect/Tensor/Transforms/PackAndUnpackPatterns.cpp:146]: Yeah... that's making comments and failure message harder. I try to make a better change. Please take a look if it is better with the suggestion.

**Review by chelini — APPROVED** [review]
Looks good to me.

**Pull request merged**